### PR TITLE
rpc: add limit for batch request and response size

### PR DESCRIPTION
**Pull request by mmsqe**
This PR adds server-side limits for JSON-RPC batch requests. Before this change, batches
were limited only by processing time. The server would pick calls from the batch and
answer them until the response timeout occurred, then stop processing the remaining batch
items.

Here, we are adding two additional limits which can be configured:

- the 'item limit': batches can have at most N items
- the 'response size limit': batches can contain at most X response bytes

These limits are optional in package rpc. In Geth, we set a default limit of 1000 items
and 25MB response size.

When a batch goes over the limit, an error response is returned to the client. However,
doing this correctly isn't always possible. In JSON-RPC, only method calls with a valid
`id` can be responded to. Since batches may also contain non-call messages or
notifications, the best effort thing we can do to report an error with the batch itself is
reporting the limit violation as an error for the first method call in the batch. If a batch is
too large, but contains only notifications and responses, the error will be reported with
a null `id`.

The RPC client was also changed so it can deal with errors resulting from too large
batches. An older client connected to the server code in this PR could get stuck
until the request timeout occurred when the batch is too large. **Upgrading to a version
of the RPC client containing this change is strongly recommended to avoid timeout issues.**

For some weird reason, when writing the original client implementation, @fjl worked off of
the assumption that responses could be distributed across batches arbitrarily. So for a
batch request containing requests `[A B C]`, the server could respond with `[A B C]` but
also with `[A B] [C]` or even `[A] [B] [C]` and it wouldn't make a difference to the
client.

So in the implementation of BatchCallContext, the client waited for all requests in the
batch individually. If the server didn't respond to some of the requests in the batch, the
client would eventually just time out (if a context was used).

With the addition of batch limits into the server, we anticipate that people will hit this
kind of error way more often. To handle this properly, the client now waits for a single
response batch and expects it to contain all responses to the requests.

### CODE REVIEW

File: rpc/errors.go [hunk 1]
```diff
@@ -61,12 +61,14 @@ const (
 	errcodeDefault                  = -32000
 	errcodeNotificationsUnsupported = -32001
 	errcodeTimeout                  = -32002
+	errcodeResponseTooLarge         = -32003
```
Comment (fjl) [inline rpc/errors.go:64]: This condition could have its own error code!

**Review by fjl — COMMENTED** [review]
This looks very good! 

I think it would be better to make these limits configurable somehow. My suggestion for that would be a method like `SetBatchLimits(length int, size int)` on `Server` and `Client`.

File: rpc/errors.go (hunk 1 above)
Comment (mmsqe) [reply rpc/errors.go:64]: yes? `-32003` is own error code

**Comment by mmsqe** [comment]
> This looks very good!
> 
> I think it would be better to make these limits configurable somehow. My suggestion for that would be a method like `SetBatchLimits(length int, size int)` on `Server` and `Client`.

Since the limit make more sense to add in server side, I only move set batch limit related change in `Server` and keep `Client` unchanged.

File: rpc/errors.go (hunk 1 above)
Comment (fjl) [reply rpc/errors.go:64]: Sorry, you're right, I didn't see the difference with `-32603` below.

Comment (mmsqe) [reply rpc/errors.go:64]: yea, these 2 [codes](https://github.com/ethereum/go-ethereum/blob/667a4088d22ada3878a500deab16dd665824a1d8/rpc/errors.go#L65-L66) are the same.

**Comment by mmsqe** [comment]
> The client can also accept requests from the server (and the server is implemented using Client) :)

May I ask how to add config to Client, which doesn't have access to this [config](https://github.com/ethereum/go-ethereum/blob/6d2ce2449949245b130bcccc97db48dedeafcebb/node/config.go#L46) right? And [newClient](https://github.com/ethereum/go-ethereum/blob/6d2ce2449949245b130bcccc97db48dedeafcebb/rpc/client.go#L224) would have breaking change by adding new params.

File: cmd/utils/flags.go [hunk 2]
```diff
@@ -786,6 +786,18 @@ var (
 		Usage:    "Allow for unprotected (non EIP155 signed) transactions to be submitted via RPC",
 		Category: flags.APICategory,
 	}
+	BatchRequestLimit = &cli.IntFlag{
+		Name:     "batch.request-limit",
```
Comment (holiman) [inline cmd/utils/flags.go:790]: IMO these flags should be somwhere in the `rpc.` namespace

**Comment by fjl** [comment]
I've changed my mind a little bit, `SetBatchLimits` on `Client` is not a good idea because it would require a connection reset (which you implemented). The limits can now be configured through DialOptions instead.

I think that there will be very few situations in which these limits need to be touched at all. The defaults should be set high enough to cover most people's needs.

One remaining thing before we can consider merging this: there should be a test for the batch item limit. This test is pretty hard to write right now because the client has no way of associating the error response with the batch. The error response has no `"id"` field. I want to make `client.BatchCall(...)` return a non-nil error when the batch has too many items. Not sure yet how to achieve this.

**Comment by mmsqe** [comment]
> The error response has no `"id"` field. I want to make `client.BatchCall(...)` return a non-nil error when the batch has too many items. Not sure yet how to achieve this behavior.

Just add the handle of msg id for batch too large case, see if it's ok

File: rpc/handler.go [hunk 3]
```diff
@@ -193,6 +185,15 @@ func (h *handler) handleBatch(msgs []*jsonrpcMessage) {
 	if len(calls) == 0 {
 		return
 	}
+
+	if len(calls) > h.batchRequestLimit && h.batchRequestLimit != 0 {
```
Comment (holiman) [inline rpc/handler.go:189]: Why was this moved -- if we want resiliency, it's better to handle batch too large before handling `non-call messages`

File: rpc/handler.go [hunk 4]
```diff
@@ -19,13 +19,14 @@ package rpc
 import (
 	"context"
 	"encoding/json"
+	"log"
```
Comment (holiman) [inline rpc/handler.go:22]: Was this an intentional or accidental change? We don't need the `log` package, we use our own

File: rpc/handler.go [hunk 5]
```diff
@@ -172,6 +176,15 @@ func (h *handler) handleBatch(msgs []*jsonrpcMessage) {
 		return
 	}
 
+	if len(msgs) > h.batchRequestLimit && h.batchRequestLimit != 0 {
+		h.startCallProc(func(cp *callProc) {
+			resp := msgs[0].errorResponse(&invalidRequestError{errMsgBatchTooLarge})
+			h.conn.writeJSON(cp.ctx, resp, true)
+		})
+		log.Println("return for batch too large")
```
Comment (holiman) [inline rpc/handler.go:184]: Either drop this or use our own `log` ( but with e.g. `log.Debug("..")` )

**Review by holiman — CHANGES_REQUESTED** [review]

File: rpc/server_test.go [hunk 6]
```diff
@@ -206,3 +209,45 @@ func TestServerBatchRequestLimit(t *testing.T) {
...
+	defer server.Stop()
+
+	client := DialInProc(server)
+	args := make([][]interface{}, 3)
+	for i := range args {
+		args[i] = make([]interface{}, 0)
+	}
+	client.BatchNotify(context.Background(), "ex_subscription", args)
+	scanner.Scan()
+	if scanner.Text() != "return for batch too large" {
```
[the remark this answers is not in the document]
Comment (mmsqe) [reply rpc/server_test.go]: Seems need more modify in client to test non-call if we don't want to compare log result.

**Comment by fjl** [comment]
@mmsqe You can leave this as-is now. I will go over it and get the PR across the finish line as soon as I have time for it.

**Review by alice22foster — CHANGES_REQUESTED** [review]
RPC Changes

**Review by alice22foster — CHANGES_REQUESTED** [review]
RPC

**Comment by fjl** [comment]
I promise I'll get back to this very soon!

**Comment by fjl** [comment]
@holiman PTAL
